pkg/splunk/enterprise/afwscheduler.go

```diff
@@ -273,12 +280,177 @@ func getPhaseInfoByPhaseType(worker *PipelineWorker, phaseType enterpriseApi.App
+	localPath := filepath.Join(splcommon.AppDownloadVolume, "downloadedApps", downloadWorker.cr.GetNamespace(), kind, downloadWorker.cr.GetName(), scope, downloadWorker.appSrcName) + "/"
+	// create the sub-directories on the volume for downloading scoped apps
+	err := createAppDownloadDir(localPath)
+	if err != nil {
+		scopedLog.Error(err, "unable to create app download directory on operator")
+	}
+	return localPath, err
+}
+
+// Download API will do the actual work of downloading apps from remote storage
+func (downloadWorker *PipelineWorker) Download(pplnPhase *PipelinePhase, s3ClientMgr S3ClientManager, activeWorkers *uint64, localPath string) {
+
+	defer func() {
+		// decrement the number of active workers
+		pplnPhase.mutex.Lock()
+		*activeWorkers--
+		pplnPhase.mutex.Unlock()
+
+		downloadWorker.isActive = false
+
+		// decrement the waiter count
+		downloadWorker.waiter.Done()
+	}()
+
+	splunkCR := downloadWorker.cr
+	appSrcName := downloadWorker.appSrcName
+	scopedLog := log.WithName("PipelineWorker.Download()").WithValues("name", splunkCR.GetName(), "namespace", splunkCR.GetNamespace(), "App name", downloadWorker.appDeployInfo.AppName, "objectHash", downloadWorker.appDeployInfo.ObjectHash)
+
+	appDeployInfo := downloadWorker.appDeployInfo
+	appName := appDeployInfo.AppName
+
+	localFile := getLocalAppFileName(localPath, appName, appDeployInfo.ObjectHash)
+	remoteFile, err := getRemoteObjectKey(splunkCR, downloadWorker.afwConfig, appSrcName, appName)
+	if err != nil {
+		scopedLog.Error(err, "unable to get remote object key", "appName", appName)
+		// increment the retry count and mark this app as download pending
+		updatePplnWorkerPhaseInfo(appDeployInfo, appDeployInfo.PhaseInfo.RetryCount+1, enterpriseApi.AppPkgDownloadPending)
+		return
+	}
+
+	// download the app from remote storage
+	err = s3ClientMgr.DownloadApp(remoteFile, localFile, appDeployInfo.ObjectHash)
+	if err != nil {
+		scopedLog.Error(err, "unable to download app", "appName", appName)
+		// increment the retry count and mark this app as download pending
+		updatePplnWorkerPhaseInfo(appDeployInfo, appDeployInfo.PhaseInfo.RetryCount+1, enterpriseApi.AppPkgDownloadPending)
+		return
+	}
+
+	// download is successfull, update the state and reset the retry count
+	updatePplnWorkerPhaseInfo(appDeployInfo, 0, enterpriseApi.AppPkgDownloadComplete)
+
+	scopedLog.Info("Finished downloading app")
+}
+
+// scheduleDownloads schedules the download workers to download app/s
+func (ppln *AppInstallPipeline) scheduleDownloads(pplnPhase *PipelinePhase, maxWorkers uint64, scheduleDownloadsWaiter *sync.WaitGroup) {
+	var activeWorkers uint64
+
+	scopedLog := log.WithName("scheduleDownloads")
+
+downloadWork:
+	for {
+		if activeWorkers <= maxWorkers {
+			select {
+			case _, channelOpen := <-ppln.sigTerm:
+				if !channelOpen {
+					scopedLog.Info("Received the termination request from the scheduler")
+					break downloadWork
+				}
```

Comment:
Can we use the msgChannel close as the exit criteria, that way the control is just in between the parent and child, to accommodate any future clean-up activity(if required) for the parent?

Reply:
Sure